src/site/antora/modules/ROOT/pages/manual/lookups.adoc

```diff
@@ -16,63 +16,298 @@
+|
+| A Java
+https://docs.oracle.com/javase/{java-target-version}/docs/api/java/util/ResourceBundle.html[resource bundle]
 
-The ContextMapLookup allows applications to store data in the Log4j
-ThreadContext Map and then retrieve the values in the Log4j
-configuration. In the example below, the application would store the
-current user's login id in the ThreadContext Map with the key "loginId".
-During initial configuration processing the first '$' will be removed.
-The PatternLayout supports interpolation with Lookups and will then
-resolve the variable for each event. Note that the pattern
-"%X\{loginId}" would achieve the same result.
+| <<DockerLookup,`docker`>>
+| log4j-docker
+| Docker container
+
+| <<EnvironmentLookup,`env`>>
+|
+| Environment variables
+
+| <<JavaLookup,`java`>>
+|
+| JVM characteristics
+
+| <<Log4jLookup,`log4j`>>
+|
+| Location of Log4j configuration file
+
+| <<MainMapLookup,`main`>>
+|
+| JVM application arguments
+
+| <<MarkerLookup,`marker`>>
+|
+| Returns `key` if a marker named `key` exists
+
+| <<SpringBootLookup,`spring`>>
+| log4j-spring-boot
+| Spring Boot 2.x environment.
+
+| <<SystemPropertiesLookup,`sys`>>
+|
+| Java system properties
+
+| <<WebLookup,`web`>>
+| log4j-jakarta-web
+| Jakarta
+https://jakarta.ee/specifications/servlet/6.0/apidocs/jakarta.servlet/jakarta/servlet/servletcontext[`ServletContext`].
+
+|===
+
+* A second group of lookups either gives results that vary in time or needs to be evaluated in the context of a log event.
+These lookups should be evaluated lazily using the `$$+{prefix:key}+` syntax.
+See
+xref:manual/configuration.adoc#lazy-property-substitution[lazy property substitution]
+for details.
++
+[#global-context-mutable-list]
+.Mutable lookups available in the global context
+[cols="1,2m,5"]
+|===
+| Prefix | Dependency | Data source
+
+| <<ContextMapLookup,`ctx`>>
+|
+| xref:manual/thread-context.adoc[]
+
+| <<DateLookup,`date`>>
+|
+| Current timestamp
+
+| <<JndiLookup,`jndi`>>
+|
+| JNDI
+
+|===
++
+[#event-context-list]
+.Lookups available in the context of a log event
+[cols="1,2m,5"]
+|===
+| Prefix | Dependency | Data source
+
+| <<ContextMapLookup,`ctx`>>
+|
+| Log event
+link:../javadoc/log4j-core/org/apache/logging/log4j/core/LogEvent.html#getContextData()[context data]
+
+| <<DateLookup,`date`>>
+|
+| Log event
+link:../javadoc/log4j-core/org/apache/logging/log4j/core/LogEvent.html#getTimeMillis()[timestamp]
+
+| <<EventLookup,`event`>>
+|
+| link:../javadoc/log4j-core/org/apache/logging/log4j/core/LogEvent.html[Log event]
+
+| <<MapLookup,`map`>>
+|
+| xref:manual/messages.adoc#MapMessage[`MapMessage`]
+
+| <<MarkerLookup,`marker`>>
+|
+| Log event
+link:../javadoc/log4j-core/org/apache/logging/log4j/core/LogEvent.html#getMarker()[marker]
+
+| <<StructuredDataLookup,`sd`>>
+|
+| xref:manual/messages.adoc#StructuredDataMessage[`StructuredDataMessage`]
+
+|===
+
+* Log4j Core also provides two lookups that can be used to modify the result of another lookup:
```

Comment:
> I think this is where my discomfort for the followed 3-way categorization has its peak. These two look out of place, but actually they are not. They seem like they are due to the grouping followed. You say for the 2nd group, not this one:

The split is an attempt to prevent expressions like `$${sys:something}` from appearing in the configuration file.
Most of the time it will just add an overhead to logging and evaluate to the same string.

> 
> > These lookups should be evaluated lazily using the `$${prefix:key}` syntax
> 
> 3rd group can also be evaluated lazily.

I chose to separate those lookups, because the way you evaluate them mostly depends on **what** you are evaluating.
I wouldn't use a constant as key, since it is easier to capitalize the key manually.

> In conclusion, I liked the 2-way classification you had earlier: lookups supporting global context and log event context. I would rather present this in a matrix composed of columns `Lookup`, `Supports global context?`, and `Supports log event context?`.

Sounds good to me.

Reply:
I moved these to the list of lookups that support evaluation in the global context: fd9f75ed4be43ac9ae53f176d816fca0e0ff1140